classes/uki-esp-image.bbclass

```diff
@@ -0,0 +1,59 @@
+# Create an ESP image that has an UKI, systemd-boot and the matching configs
+
+# Optional subfolder, dependant on where the ESP partition gets mounted
+# intended to only have a leading slash, no trailing slash e.g. '/EFI', or just empty, ''
+ESPFOLDER ?= "/EFI"
+
+# Optional initrd, filname inside ${DEPLOY_DIR_IMAGE}
+ESPINITRD ?= ""
+
+# Title as shown on the console
+UKI_TITLE ?= "${DISTRO_NAME} ${DISTRO_VERSION} - ${MACHINE}"
+
+# Kernel CMDLINE
+UKI_CMDLINE = " "
+UKI_CMDLINE[export] = "1"
+
+do_ukiesp() {
+	mkdir -p ${IMAGE_ROOTFS}${ESPFOLDER}/loader/entries
+
+	# Copy over files from deploy into the rootfs
+	install -m 0755 ${DEPLOY_DIR_IMAGE}/${UKI_FILENAME} ${IMAGE_ROOTFS}${ESPFOLDER}
+	install -m 0644 ${DEPLOY_DIR_IMAGE}/${KERNEL_DEVICETREE} ${IMAGE_ROOTFS}${ESPFOLDER}
+
+	# Add global settings to systemd-boot config
+	echo "timeout 5" > ${IMAGE_ROOTFS}${ESPFOLDER}/loader/loader.conf
+
+	# copy over kernel and optional initrd and dtb
+	echo "title ${UKI_TITLE}" > ${IMAGE_ROOTFS}${ESPFOLDER}/loader/entries/boot.conf
+	echo "linux /${UKI_FILENAME}" >> ${IMAGE_ROOTFS}${ESPFOLDER}/loader/entries/boot.conf
+
+	# Assemble CMDLINE
+	# FIXME: don't use QCOM_* variables
+	if [ -n "${QCOM_BOOTIMG_ROOTFS}" ]; then
+	    UKI_CMDLINE="$UKI_CMDLINE root=${QCOM_BOOTIMG_ROOTFS} rw rootwait"
+	fi
+	if [ -n "${SERIAL_CONSOLES}" ]; then
+	    tmp="${SERIAL_CONSOLES}"
+	    for entry in $tmp ; do
+		baudrate=`echo $entry | sed 's/\;.*//'`
+		tty=`echo $entry | sed -e 's/^[0-9]*\;//' -e 's/\;.*//'`
+		console="$tty","$baudrate"n8
+		UKI_CMDLINE="$UKI_CMDLINE console=$console"
+	    done
+	fi
+	if [ -n "${KERNEL_CMDLINE_EXTRA}" ]; then
+	    UKI_CMDLINE="$UKI_CMDLINE ${KERNEL_CMDLINE_EXTRA}"
+	fi
+	echo "options $UKI_CMDLINE" >> ${IMAGE_ROOTFS}${ESPFOLDER}/loader/entries/boot.conf
+
+	if [ -n "${ESPINITRD}" ] ; then
+		install -m 0644 ${DEPLOY_DIR_IMAGE}/${ESPINITRD} ${IMAGE_ROOTFS}${ESPFOLDER}
+		echo "initrd /${ESPINITRD}" >> ${IMAGE_ROOTFS}${ESPFOLDER}/loader/entries/boot.conf
+	fi
+	echo "devicetree /${KERNEL_DEVICETREE}" >> ${IMAGE_ROOTFS}${ESPFOLDER}/loader/entries/boot.conf
```

Comment:
I generally feel that specifying extra DT, initrd or kernel command line defeats the purpose of the UKI in some way. Ideally it should be a single and the only file that is being loaded by the bootloader.  For example, `loader.conf` can not be signed.

Reply:
I agree, but the aim for this PR is to match what the WIC plugin does, so we have an actual, bootable example to show why this is unsuited for the use cases we want to support. Having said that, this should be the highest quality implementation that matches that plugin!

Reply:
my 2c would be to drop the `initrd` and `devicetree` from the beginning (and maybe the commandline as it is so unobvios). In the end, the major obstacle with WIC is in the image format, not in some other property.

Reply:
Yeah, I think it is better to just support type 2 entries instead (https://uapi-group.org/specifications/specs/boot_loader_specification/#type-2-efi-unified-kernel-images), with everything included as part of the UKI.